src/types/utils.cpp

```diff
@@ -378,71 +101,239 @@ til::color Utils::ColorFromHexString(const std::string_view str)
-bool Utils::IsValidHandle(const HANDLE handle) noexcept
+// - An optional color which contains value if a color was successfully parsed
+std::optional<til::color> Utils::ColorFromXTermColor(const std::wstring_view string) noexcept
 {
-    return handle != nullptr && handle != INVALID_HANDLE_VALUE;
-}
+    std::optional<til::color> color = ColorFromXParseColorSpec(string);
+    if (color.has_value())
+    {
+        return color.value();
+    }
 
-// Function Description:
-// - Fill the first 16 entries of a given color table with the Campbell color
-//   scheme, in the ANSI/VT RGB order.
-// Arguments:
-// - table: a color table with at least 16 entries
-// Return Value:
-// - <none>, throws if the table has less that 16 entries
-void Utils::InitializeCampbellColorTable(const gsl::span<COLORREF> table)
-{
-    THROW_HR_IF(E_INVALIDARG, table.size() < 16);
+    color = ColorFromXOrgAppColorName(string);
+    if (color.has_value())
+    {
+        return color.value();
+    }
 
-    std::copy(campbellColorTable.begin(), campbellColorTable.end(), table.begin());
+    return std::nullopt;
 }
 
-// Function Description:
-// - Fill the first 16 entries of a given color table with the Campbell color
-//   scheme, in the Windows BGR order.
+// Routine Description:
+// - Given a color spec string, attempts to parse the color that's encoded.
+//
+//   Based on the XParseColor documentation, the supported specs currently are the following:
+//      spec1: a color in the following format:
+//          "rgb:<red>/<green>/<blue>"
+//      spec2: a color in the following format:
+//          "#<red><green><blue>"
+//
+//   In both specs, <color> is a value contains up to 4 hex digits, upper or lower case.
 // Arguments:
-// - table: a color table with at least 16 entries
+// - string - The string containing the color spec string to parse.
 // Return Value:
-// - <none>, throws if the table has less that 16 entries
-void Utils::InitializeCampbellColorTableForConhost(const gsl::span<COLORREF> table)
+// - An optional color which contains value if a color was successfully parsed
+std::optional<til::color> Utils::ColorFromXParseColorSpec(const std::wstring_view string) noexcept
+try
 {
-    THROW_HR_IF(E_INVALIDARG, table.size() < 16);
-    InitializeCampbellColorTable(table);
-    SwapANSIColorOrderForConhost(table);
+    bool foundXParseColorSpec = false;
+    bool foundValidColorSpec = false;
+
+    bool isSharpSignFormat = false;
+    size_t rgbHexDigitCount = 0;
+    std::array<unsigned int, 3> colorValues = { 0 };
+    std::array<unsigned int, 3> parameterValues = { 0 };
+    const auto stringSize = string.size();
+
+    // First we look for "rgb:"
+    // Other colorspaces are theoretically possible, but we don't support them.
+    auto curr = string.cbegin();
+    if (stringSize > 4)
+    {
+        auto prefix = std::wstring(string.substr(0, 4));
+        std::transform(prefix.begin(), prefix.end(), prefix.begin(), ::towlower);

```

Comment:
I thought about the `towlower` behaviour under different locales. I do not have enough knowledge in this area, though. If there's a locale that has `:` as lower form of a character, then this should be fixed.

Reply:
The XOrg code uses `_XcmsCopyISOLatin1Lowered` in `_XcmsParseColorString`, so we only need to `tolower` in ISO Latin-1, irrespective of the current locale.